src/utils/sepana.ts

```diff
@@ -1,10 +1,173 @@
   j: Json<SepanaProject>,
 ): SepanaProject => ({
   ...j,
   ...parseBigNumberKeyVals(j, ['currentBalance', 'totalPaid', 'trendingScore']),
 })
+
+export function getChangedSubgraphProjects({
+  subgraphProjects,
+  sepanaProjects,
+  retryIpfs,
+}: {
+  subgraphProjects: Json<Pick<Project, SGSepanaCompareKey>>[]
+  sepanaProjects: Record<string, Json<SepanaProject>>
+  retryIpfs?: boolean
+}) {
+  const idsOfNewProjects = new Set<string>([])
+
+  const updatedProperties: {
+    [id: string]: {
+      key: string
+      oldVal: string | undefined | null
+      newVal: string | undefined | null
+    }[]
+  } = {}
+
+  let retryMetadataCount = 0
+
+  const changedSubgraphProjects = subgraphProjects.filter(subgraphProject => {
+    const id = subgraphProject.id
+
+    const sepanaProject = sepanaProjects[subgraphProject.id]
+
+    if (!sepanaProject) {
+      idsOfNewProjects.add(id)
+      return true
+    }
+
+    if (sepanaProject._version !== CURRENT_VERSION) {
+      return true
+    }
+
+    const { _hasUnresolvedMetadata, _metadataRetriesLeft } = sepanaProject
+
+    if (
+      retryIpfs &&
+      _hasUnresolvedMetadata &&
+      (_metadataRetriesLeft || _metadataRetriesLeft === undefined)
+    ) {
+      retryMetadataCount++
+      return true
+    }
+
+    // Deep compare Subgraph project vs. Sepana project and find any discrepancies
+    const propertiesToUpdate = sgSepanaCompareKeys.filter(k => {
+      const oldVal = sepanaProject[k]
+      const newVal = subgraphProject[k]
+
+      // Store a record of properties that need updating
+      if (oldVal !== newVal) {
+        updatedProperties[id] = [
+          ...(updatedProperties[id] ?? []),
+          {
+            key: k,
+            oldVal: oldVal?.toString(),
+            newVal: newVal?.toString(),
+          },
+        ]
+        return true
+      }
+
+      return false
+    })
+
+    // Return true if any properties are out of date
+    return propertiesToUpdate.length
+  })
+
+  return {
+    changedSubgraphProjects,
+    updatedProperties,
+    retryMetadataCount,
+    idsOfNewProjects,
+  }
+}
+
+export async function tryResolveMetadata({
+  subgraphProject,
+  _metadataRetriesLeft,
+  _hasUnresolvedMetadata,
+}: {
+  subgraphProject: Json<Pick<Project, SGSepanaCompareKey>>
+} & Partial<
+  Pick<SepanaProject, '_hasUnresolvedMetadata' | '_metadataRetriesLeft'>
+>) {
+  const { metadataUri } = subgraphProject
+
+  // if metadataUri is missing or invalid, or no retries remaining for unresolved metadata
+  if (
+    !metadataUri ||
+    !isIpfsCID(metadataUri) ||
+    (_hasUnresolvedMetadata && _metadataRetriesLeft === 0)
+  ) {
+    return {
+      project: {
+        ...subgraphProject,
+        _hasUnresolvedMetadata: true,
+        _metadataRetriesLeft: 0,
+      } as Json<SepanaProject>,
+    }
+  }
+
+  try {
+    const {
+      data: { name, description, logoUri },
+    } = await infuraApi.get<AnyProjectMetadata>(openIpfsUrl(metadataUri), {
```

Comment:
**suggestion** could potentially use `ipfsGetWithFallback` here.

Reply:
was originally doing this, but that was using the pinata url at the time which was failing on stuff that wasn't pinned via our gateway. will clean this up at some point but gonna leave for this PR